pkg/datastore/database.go

```diff
@@ -0,0 +1,771 @@
+	tx.Commit()
+	if tx.RowsAffected == 0 {
+		err = ErrRecordNotFound.WithMap(map[ErrorContextKey]string{PRIMARY_KEY: fmt.Sprintf("%+v", record.GetId())})
+		db.logger.Error(err)
+		return err
+	}
+	if tx.Error != nil {
+		err = ErrExecutingSqlStmt.Wrap(tx.Error)
+		db.logger.Error(err)
+		return err
+	}
+	return nil
+}
+
+// Finds all records in a DB table.
+// records must be a pointer to a slice of structs and will be modified in-place.
+func (db *relationalDb) FindAll(ctx context.Context, records interface{}) error {
+	if reflect.TypeOf(records).Kind() != reflect.Ptr || reflect.TypeOf(records).Elem().Kind() != reflect.Slice {
+		errMsg := "\"records\" argument has to be a pointer to a slice of structs implementing \"Record\" interface"
+		db.logger.Error(errMsg)
+		err := ErrNotPtrToStructSlice.Wrap(fmt.Errorf(errMsg))
+		return err
+	}
+
+	tableName := GetTableNameFromSlice(records)
+	return db.FindAllInTable(ctx, tableName, records)
+}
+
+// Finds all records in DB table tableName.
+// records must be a pointer to a slice of structs and will be modified in-place.
+func (db *relationalDb) FindAllInTable(ctx context.Context, tableName string, records interface{}) error {
+	record := GetRecordInstanceFromSlice(records)
+	return db.FindWithFilterInTable(ctx, tableName, record, records)
+}
+
+// Finds multiple records in a DB table.
+// If record argument is non-empty, uses the non-empty fields as criteria in a query.
+// records must be a pointer to a slice of structs and will be modified in-place.
+func (db *relationalDb) FindWithFilter(ctx context.Context, record Record, records interface{}) error {
+	return db.FindWithFilterInTable(ctx, GetTableName(record), record, records)
+}
+
+// Finds multiple records in DB table tableName.
+// If record argument is non-empty, uses the non-empty fields as criteria in a query.
+// records must be a pointer to a slice of structs and will be modified in-place.
+func (db *relationalDb) FindWithFilterInTable(ctx context.Context, tableName string, record Record, records interface{}) (err error) {
+	if reflect.TypeOf(records).Kind() != reflect.Ptr || reflect.TypeOf(records).Elem().Kind() != reflect.Slice {
+		return ErrNotPtrToStruct.WithValue(TYPE, TypeName(records))
+	}
+
+	tx, err := db.getDBTransaction(ctx, tableName, record)
+	if err != nil {
+		return err
+	}
+
+	tx.Where(record).Find(records) // FILTER by record too
+	tx.Commit()
+	if tx.Error != nil {
+		err = ErrExecutingSqlStmt.Wrap(tx.Error)
+		db.logger.Errorln(err)
+		return err
+	}
+	return nil
+}
+
+/*
+Inserts a record into a DB table.
+*/
+func (db *relationalDb) Insert(ctx context.Context, record Record) (rowsAffected int64, err error) {
+	return db.InsertInTable(ctx, GetTableName(record), record)
+}
+
+/*
+Inserts a record into DB table tableName.
+*/
+func (db *relationalDb) InsertInTable(ctx context.Context, tableName string, record Record) (rowsAffected int64, err error) {
+	tx, err := db.GetDBTransaction(ctx, tableName, record)
```

Comment:
@krishnamiriyala, just to point it out, `DataStore`'s Insert, Update, Upsert, Delete methods used to accept pointers to structs **and** structs, because, unlike the Find method, they didn't have to modify the struct passed to it. So, passing the copy of the struct instead of a pointer worked too. `GetDBTransaction()` will return an error if `record` is not a pointer, so this will be a backward-incompatible change. Not saying we should allow structs - it is inefficient - but just wanted to point it out

Reply:
Yes all the methods are restricted to PtrToStruct after this change